cpp/ql/test/library-tests/dataflow/dataflow-tests/test_self_argument_flow.ql

```diff
@@ -0,0 +1,33 @@
+import cpp
+import semmle.code.cpp.dataflow.new.DataFlow
+import TestUtilities.InlineExpectationsTest
+
+module TestConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) {
+    source.getLocation().getFile().getBaseName() = "self_argument_flow.cpp" and
+    source.asDefiningArgument() =
+      any(Call call | call.getTarget().hasName("acquire")).getAnArgument()
+  }
+
+  predicate isSink(DataFlow::Node sink) {
+    sink.asIndirectArgument() = any(Call call | call.getTarget().hasName("acquire")).getAnArgument()
+  }
```

Comment:
What does this test show exactly?

Reply:
This is an example of loopy dataflow that I was observing in DCA. When we have a call such as the following:
```cpp
void write_to(int* p);

struct S { int* p };

void test(S* s) {
  write_to(s->p);
}
```
we would not expect flow to come out of `write_to` and then flow back into the same call to `write_to` again. This is a reduced test that represents such invalid flow (and you can see that we no longer have this flow in 99cc4171f80d6a14a5378545cb1e8759ff3b5a55).